ozaria/site/components/play/PageIntroLevel/index.vue

```diff
@@ -0,0 +1,134 @@
+<script>
+import api from 'core/api'
+import interactivesComponent from '../../interactive/PageInteractive'
+import cinematicsComponent from '../../cinematic/PageCinematic'
+import { defaultCodeLanguage } from 'ozaria/site/common/ozariaUtils'
+
+module.exports = Vue.extend({
+  props: {
+    introLevelIdOrSlug: {
+      type: String,
+      required: true,
+      default: ''
+    },
+    courseInstanceId: {
+      type: String
+    },
+    codeLanguage: {
+      type: String
+    },
+    courseId: {
+      type: String
+    },
+  },
+  data: () => ({
+    introLevelData: {},
+    introLevelSession: {},
+    introContent: [],
+    currentContent: {},
+    currentIndex: 0,
+    language: ''
+  }),
+  components: {
+    'interactives-component': interactivesComponent,
+    'cinematics-component': cinematicsComponent
+    // TODO add when ready
+    // 'cutscene-video-component': cutsceneVideoComponent,
+    // 'avatar-selection-screen': avatarSelectionScreen
+  },
+  async created () {
+    if (!me.hasIntroLevelAccess()) {
+      alert('You must be logged in as an admin to use this page.')
+      return application.router.navigate('/', { trigger: true })
+    }
+    try {
+      this.introLevelData = await api.levels.getByIdOrSlug(this.introLevelIdOrSlug)
+      if (me.isSessionless()) {
+        this.language = this.codeLanguage || defaultCodeLanguage
+      } else {
+        this.introLevelSession = await api.levels.upsertSession(this.introLevelIdOrSlug, {courseInstanceId: this.courseInstanceId})
+        this.language = this.introLevelSession.codeLanguage
+      }
+      
+      const content = this.introLevelData.introContent
+      if (_.isArray(content)) {
+        this.introContent = content
+      } else if (_.isObject(content)) {
+        if (!content[this.language]) {
+          console.error(`Intro content for language ${this.language} not found`)
+          noty({text: 'Invalid intro content', type: 'error', timeout:2000})
+          return
+        }
+        this.introContent = content[this.language]
+      } else {
+        console.error('Invalid intro content, it should be an array or an object')
+        noty({text: 'Invalid intro content', type: 'error', timeout: 2000})
+        return
+      }
+    } catch (err) {
+      console.error('Error in creating data for intro level', err)
+      noty({text: 'Error in creating data for intro level', type: 'error', timeout: 2000})
+      return
+    }
+    // Assign first content in the sequence to this.currentContent
+    this.currentIndex = 0
+    this.currentContent = this.introContent[this.currentIndex] 
+  },
+  methods: {
+    onContentCompleted: async function () {
```

Comment:
How do we intended to handle failures within sub components?

Reply:
What kind of failures do you refer to here, I think the subcomponents should be handling their own failures/successes. Only when they are complete and ready to move on, this method will be triggered.

Reply:
I wonder if there is ever a situation where a component crashes and blocks progress?

Maybe it is completely on the sub component side. It may also depend if it's possible.
Regarding the Cinematic view. If the cinematic crashes and starts throwing errors everywhere. It should still allow completion.

However if the cinematic doesn't load properly (intermittent 504), it may be good if it can emit or throw and error and have the intro retry/rebuild it.

Not blocking.

Reply:
If a component crashes it'll block progress for the current page load but hopefully won't corrupt data such that the student is now stuck.  The question is how do we want to handle any failures in the sub components?  Do we want to display a generic error screen that says "refresh the page and try again"?  Do we want to just let the component hang there?

The sub components should definitely handle their own errors (as they do now with noty).  We may want to communicate error status somewhere so that we can offer a consistent UX vs randomly stuck pages.

Reply:
504s are a great example.  If something fails to load or something in the sub component fails to initialize what do we want to do?

Reply:
Loading state is a similar thing - we'll likely want a single loading indicator that is controlled by the top level component with state communicated by the sub components.

Alternatively we can use a single loading component within each sub component.  Either way something we'll need to think about at the design level probably along with how we want to handle error states across the lifecycle of an intro level.

Reply:
Great points @AndrewJakubowicz  and @jmif, thanks for bringing these up!
>  If the user refreshes an intro will it restart them from the very start?

Yes, it would restart from the beginning(however this may change in future). I agree with having some retry mechanism so that they do not have to refresh whenever this occurs / or atleast communicate the error across a consistent UX. 

Thanks Joe for opening this discussion about how we should handle error states across the lifecycle of an intro level. I like the idea about a loading indicator within each sub component since intro level doesnt load all the sub components at the beginning.
I am also open to other suggestions and adding it as an optimization task in the wrap-up tasks list for now.

Reply:
Lets add a task to track and follow up later

Reply:
Yep added here: https://app.asana.com/0/1117972742375106/1125361899036790